docs/03_server/04_event-handler/03_advanced.md

```diff
@@ -241,110 +251,247 @@ eventHandler {
 
-- `userName` - a string property containing the user name who triggered the event (insert, accept, reject or cancel).
-- `db` - an RxDb property so you can access the database layer and do appropriate checks.
-- `pendingApproval` - the pending approval record stored in the database (only available in accept, reject or cancel events).
+```kotlin
+pendingApproval {
+    accept {
+        val userAttributes = entityDb.get(UserAttributes.byName(userName))
+        userAttributes?.accessType == AccessType.INTERNAL
+    }
+}
+```
+
+### Configuring allowed approvers: Advanced
+
+You might have noticed that the original type-safe event message types are lost inside the *-approval.kts* file, as the content of `eventMessage` inside `APPROVAL` table (and also inside `PendingApprovalInsert`) is a serialized JSON string. It is possible to deserialize the original type-safe objects by using the `selectPredicate` method combined with multiple `onEvent` predicates. These methods are available in all of the `pendingApproval` code blocks: `insert`, `accept`, `cancel` and `reject`.
 
-With this xml configuration, you can look up the user's rights in the database and return `true` only if the necessary rights are in place. For example, if your system has the concept of internal and external users and you only want to allow internal users to accept pending events, then you could check your custom user "ACCESS_TYPE" field as follows:
+- `selectPredicate` is a function that accepts an indeterminate number of functions returning a boolean value, as well as a mandatory `default` function to handle messages that do not fall into any defined category. The `default` function will provide a [GenesisSet](../../03_server/09_inter-process-messages/02_genesisset.md) object with the contents of the original message payload.
+- `onEvent` works very similarly to any other GPAL [event handler definition](../../server/event-handler/basics.md#adding-a-name). It allows us to treat the incoming message in the same way we would have done it within the original event handler, but each function must return a boolean expression.
 
-```xml
-        ...
-        <acceptPendingApproval>
-            <![CDATA[
-               def searchRecord = new DbRecord("USER_ATTRIBUTES")
-               def userAttributes = db.get(searchRecord, "USER_ATTRIBUTES_BY_USER_NAME")
-               userAttributes?.getString("ACCESS_TYPE") == "INTERNAL"
-            ]]>
-        </acceptPendingApproval>
-        ...
+
+Please see the example below:
+
+```kotlin
+pendingApproval {
+    accept {
+        selectPredicate(        
+            onEvent<TradeInsert>("TRADE_INSERT") { event ->
+                val tradeInsert = event.details
+                // Allows acceptance of pending approval events for EVENT_TRADE_INSERT, only if the trade is for TSLA
+                tradeInsert.symbol == "TSLA"
```

Comment:
this logic is a bit random. should we move to mak up rights codes like InsertTrade, AmendTrade and DeleteTrade

Reply:
will do. I just wanted to show the fact that you can access the original values from the event. I will try to create a new example that does something with the values as well as the codes.